### DISCUSSION TITLE
Next.js ( API Routes ) + socket.io Real-time chat

### DISCUSSION THREAD

**Discussion by jeongminsang** [vercel/next.js#56002] [Help]
## Feature request

I created a Real-time chat through Next.js API Routes.
However, when I deploy it with [Vercel, it doesn't work properly because Vercel's Serverless Functions don't support WebSocket.](https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections) 
What is the best solution for this? 
I think I can deploy using Vercel + AWS (API gateway + lambda), but it seems too difficult.
If possible, I would also like to know how you implemented the deployment using Vercel + AWS.

Below is the client and server code.

```ts
// pages/index.tsx
  const [message, setMessage] = useState("");
  const [messageList, setMessageList] = useState<any>([]);

  const socket = io("http://localhost:3000/", { path: "/api/socketio" });

  useEffect(() => {
    socket.on("connect", () => {
      console.log("socket server connected.");
    });
    socket.on("disconnect", () => {
      console.log("socket server disconnected.");
    });
  }, []);

  useEffect(() => {
    socket.on("receivemsg", (data) => {
      setMessageList([...messageList, data]);
    });

  }, [messageList]);

  const handleSend = () => {
    socket.emit("sendmsg", message);
    console.log(message);
    setMessage("");
  };
```
```ts
// pages/api/socketio.ts
import { Server } from "socket.io";

const socketioHandler = (req: any, res: any) => {
  if (!res.socket.server.io) {
    const io = new Server(res.socket.server, { path: "/api/socketio" });

    io.on("connection", (socket) => {
      console.log("User connected succesfully");

      socket.on("hello", (arg, callback) => {
        callback(arg);
      });

      socket.on("sendmsg", (msg) => {
        io.emit("receivemsg", msg);
      });
      socket.on("disconnect", () => {
        console.log(`Socket ${socket.id} disconnected.`);
      });
    });

    res.socket.server.io = io;
  } else {
    console.log("socket.io already running");
  }
  res.end();
};

export default socketioHandler;
```

**Comment by icyJoseph** [comment]
Well...

If you must use WebSockets, there are Vercel supported alternatives, for "live apps", using Pusher among others:

- https://vercel.com/guides/publish-and-subscribe-to-realtime-data-on-vercel#providers
- https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections

 For transparency sake, there are also alternatives to Vercel, if you must definitely use WebSockets:

- https://render.com/pricing (I think they have WS protocol on their free tier)
- https://railway.app/pricing (can't find it in their docs)

I haven't had time to try either though.
